core/src/data_sources/qdrant.rs

```diff
@@ -0,0 +1,74 @@
+use anyhow::{anyhow, Result};
+use std::collections::HashMap;
+use std::sync::Arc;
+
+use parking_lot::Mutex;
+use qdrant_client::prelude::{QdrantClient, QdrantClientConfig};
+use serde::{Deserialize, Serialize};
+
+#[derive(Debug, Clone, Copy, Serialize, PartialEq, Deserialize, Eq, Hash)]
+pub enum QdrantCluster {
+    #[serde(rename = "main-0")]
+    Main0,
+    //#[serde(rename = "dedicated-0")]
+    //Dedicated0,
+}
+
+static QDRANT_CLUSTER_VARIANTS: &[QdrantCluster] = &[QdrantCluster::Main0];
+
+pub fn env_var_prefix_for_cluster(cluster: QdrantCluster) -> &'static str {
+    match cluster {
+        QdrantCluster::Main0 => "QDRANT_MAIN_0",
+        // QDrantCluster::Dedicated0 => "QDRANT_DEDICATED_0",
+    }
+}
+
+#[derive(Clone)]
+pub struct QdrantClients {
+    clients: Arc<Mutex<HashMap<QdrantCluster, Arc<QdrantClient>>>>,
+}
+
+impl QdrantClients {
+    async fn qdrant_client(cluster: QdrantCluster) -> Result<QdrantClient> {
+        let url_var = format!("{}_URL", env_var_prefix_for_cluster(cluster));
+        let api_key_var = format!("{}_API_KEY", env_var_prefix_for_cluster(cluster));
+
+        match std::env::var(url_var.clone()) {
+            Ok(url) => {
+                let mut config = QdrantClientConfig::from_url(&url);
+                match std::env::var(api_key_var.clone()) {
+                    Ok(api_key) => {
+                        config.set_api_key(&api_key);
+                        QdrantClient::new(Some(config))
+                    }
+                    Err(_) => Err(anyhow!("{} is not set", api_key_var))?,
+                }
+            }
+            Err(_) => Err(anyhow!("{} is not set", url_var))?,
+        }
+    }
+
+    pub async fn build() -> Result<Self> {
+        let clients = futures::future::try_join_all(QDRANT_CLUSTER_VARIANTS.into_iter().map(
+            |cluster| async move {
+                let client = Self::qdrant_client(*cluster).await?;
+                Ok::<_, anyhow::Error>((*cluster, Arc::new(client)))
+            },
+        ))
+        .await?
+        .into_iter()
+        .collect::<HashMap<_, _>>();
+
+        Ok(Self {
+            clients: Arc::new(Mutex::new(clients)),
+        })
+    }
+
+    pub fn get(&self, cluster: QdrantCluster) -> Arc<QdrantClient> {
+        let clients = self.clients.lock();
```

Comment:
Would the lock() return a Result here? in that case is it worth checking for err? (not saying it should, asking)
Also wondering as a rust noob -> the lock is released automatically  after the function finishes?

Reply:
lock does not error

Reply:
otherwise the type checking would have failed here